### ISSUE TITLE
Keras STFT Layers for Speech and Music : loading for training

### ISSUE THREAD

**Issue by mpariente** [keunwoochoi/keras_STFT_layer#1]
Dear Keunwoo Choi,

My name is Manuel Pariente, I'm a student in a Cognitive Science Master in Paris (Cogmaster, ENS Ulm) and I want to implement a NN which performs source separation in audio.

I am getting interested in your Keras STFT Layers because indeed storing all the spectrograms is not very efficient but I wonder how I could apply these layers in my case.
In fact, I won't have enough space to load my training data all at once if they are spectrograms but it would be the same for audio files to be then transformed into spectrograms within the NN.

So I wonder, if you would have a large dataset of speech or musical mixtures, how would you design your training procedure in terms of memory management? Would you load "on line" and erase the data as soon as the batch computation is over? Would you make sub-epochs until you cover all your training data? Or would you have a better solution that I obviously didn't think about?

It's true that storing the audio files and the spectrograms is not efficient in a static memory perspective and from this point, the STFT layers that you designed are really helpul, but I have trouble seeing how they would facilitate training procedures.

I hope you can shed some light on my issues and thank you in advance for it,
Best,

**Comment by keunwoochoi** [comment]
Basically what you're concerning is also a general deep learning issue. But it is important in particular here as one of the goals of using `keras_stft` layer is to reduce the burden of pre-processing. 

Loading all of the data at once may take too long. One quick and effective way is to use `model.fit_generator()` since 

> Fits the model on data generated batch-by-batch by a Python generator. **The generator is run in parallel to the model, for efficiency.**

. Then another question we have is if we could keep it as compressed format such as `mp3`. The answer would be **No** in many cases. Decoding mp3 is very slow. So I stored audio signals after decoding and down-sampling. 
I stored each file as `.npy` and created a generator to load them, but this can changed. 

An alternative is, if the length of audio sample is all identical, to store them in using HDF after shuffling.

**Comment by mpariente** [comment]
Hey, thanks a lot for your answer, the model.fit_generator() tip is really helpul, I will try it out! 

So you downsample until SR if you're interested in frequencies until SR/2 and you save as .npy . 

My concern is more the speed of the NN than the static memory taken by the data before training : 
If I would calculate spectrograms in Matlab and then use model.fit_generator() to load them in parallel, would the training procedure be much slower than with the STFT layers?  (Without counting the preprocessing) 

Thank you in advance

**Comment by keunwoochoi** [comment]
How could you make Matlab and python code communicate? To do it relatively easily, computing spectrograms should be also done in Python I guess.

**Comment by mpariente** [comment]
Until now, my spectrograms were calculated in matlab and loaded in python with `scipy.io.loadmat` and it's not that slow but probably using STFT would be better!
